rpm/0069-sailfishos-webrtc-Enable-GMP-for-encoding.-JB-53982.patch

```diff
@@ -0,0 +1,477 @@
+   // 3* is because we're using 3 buffers per frame for i420 data for now.
+-  if ((NumInUse(GMPSharedMem::kGMPFrameData) >
+-       3 * GMPSharedMem::kGMPBufLimit) ||
++  if ((NumInUse(GMPSharedMem::kGMPFrameData) > 3 * GMPSharedMem::kGMPBufLimit) ||
+       (NumInUse(GMPSharedMem::kGMPEncodedData) > GMPSharedMem::kGMPBufLimit)) {
++
+     LOGE(
+         ("GMPVideoDecoderParent[%p]::Decode() ERROR; shmem buffer limit hit "
+          "frame=%d encoded=%d",
+diff --git a/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.cpp b/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.cpp
+index b5f39bc3b7d4..287ccf3a7862 100644
+--- a/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.cpp
++++ b/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.cpp
+@@ -11,8 +11,8 @@ WebrtcVideoEncoder* GmpVideoCodec::CreateEncoder() {
+   return new WebrtcVideoEncoderProxy();
+ }
+ 
+-WebrtcVideoDecoder* GmpVideoCodec::CreateDecoder() {
+-  return new WebrtcVideoDecoderProxy();
++WebrtcVideoDecoder* GmpVideoCodec::CreateDecoder(webrtc::VideoCodecType type) {
++  return new WebrtcVideoDecoderProxy(type);
+ }
+ 
+ }  // namespace mozilla
+diff --git a/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.h b/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.h
+index 318a891d0678..0fcaac43f364 100644
+--- a/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.h
++++ b/media/webrtc/signaling/src/media-conduit/GmpVideoCodec.h
+@@ -11,7 +11,7 @@ namespace mozilla {
+ class GmpVideoCodec {
+  public:
+   static WebrtcVideoEncoder* CreateEncoder();
+-  static WebrtcVideoDecoder* CreateDecoder();
++  static WebrtcVideoDecoder* CreateDecoder(webrtc::VideoCodecType type);
+ };
+ 
+ }  // namespace mozilla
+diff --git a/media/webrtc/signaling/src/media-conduit/VideoConduit.cpp b/media/webrtc/signaling/src/media-conduit/VideoConduit.cpp
+index c8dddccec26b..e74f1061ce28 100644
+--- a/media/webrtc/signaling/src/media-conduit/VideoConduit.cpp
++++ b/media/webrtc/signaling/src/media-conduit/VideoConduit.cpp
+@@ -1501,15 +1501,37 @@ webrtc::VideoDecoder* WebrtcVideoConduit::CreateDecoder(
+     return decoder;
+   }
+ 
+-  switch (aType) {
++  // Attempt to create a GMP decoder.
++  {
++    nsCString tag;
++
++    switch (aType) {
+     case webrtc::VideoCodecType::kVideoCodecH264:
+-      // get an external decoder
+-      decoder = GmpVideoCodec::CreateDecoder();
++      tag = NS_LITERAL_CSTRING("h264");
++      break;
++    case webrtc::VideoCodecType::kVideoCodecVP8:
++      tag = NS_LITERAL_CSTRING("vp8");
++      break;
++    case webrtc::VideoCodecType::kVideoCodecVP9:
++      tag = NS_LITERAL_CSTRING("vp9");
++      break;
++    default:
++      return nullptr;
++    }
++
++    if (HaveGMPFor(NS_LITERAL_CSTRING(GMP_API_VIDEO_DECODER), { tag })) {
++      decoder = GmpVideoCodec::CreateDecoder(aType);
+       if (decoder) {
+         mRecvCodecPlugin = static_cast<WebrtcVideoDecoder*>(decoder);
+       }
+-      break;
++      return decoder;
```

Comment:
I did this on purpose. If `HaveGMPFor()` reported codec availability, then null indicates an error when creating a GMP codec. Otherwise it may hide problems. Do you think it's wrong?

Reply:
That makes perfect sense and I was just checking. Thanks for the explanation.